bigquery/tests/unit/test__helpers.py

```diff
@@ -1481,6 +1528,81 @@ def test_to_api_repr_w_nested_struct(self):
         self.assertEqual(param.to_api_repr(), EXPECTED)
 
 
+class Test__query_param_from_api_repr(unittest.TestCase):
```

Comment:
Needs a tests for dates. I'm getting a system test failure with this PR

```
_________________________________________________________________________________ TestBigQuery.test_dbapi_w_query_parameters __________________________________________________________________________________
Traceback (most recent call last):
  File "/Users/swast/src/google-cloud-python/bigquery/tests/system.py", line 1048, in test_dbapi_w_query_parameters
    Config.CURSOR.execute(example['sql'], example['query_parameters'])
  File "/Users/swast/src/google-cloud-python/bigquery/google/cloud/bigquery/dbapi/cursor.py", line 158, in execute
    query_results = query_job.result()
  File "/Users/swast/src/google-cloud-python/bigquery/google/cloud/bigquery/job.py", line 1404, in result
    super(QueryJob, self).result(timeout=timeout)
  File "/Users/swast/src/google-cloud-python/bigquery/google/cloud/bigquery/job.py", line 515, in result
    self.begin()
  File "/Users/swast/src/google-cloud-python/bigquery/google/cloud/bigquery/job.py", line 398, in begin
    self._set_properties(api_response)
  File "/Users/swast/src/google-cloud-python/bigquery/google/cloud/bigquery/job.py", line 348, in _set_properties
    self._copy_configuration_properties(configuration)
  File "/Users/swast/src/google-cloud-python/bigquery/google/cloud/bigquery/job.py", line 1358, in _copy_configuration_properties
    for mapping in configuration.get(self._QUERY_PARAMETERS_KEY, ())
  File "/Users/swast/src/google-cloud-python/bigquery/google/cloud/bigquery/job.py", line 1358, in <listcomp>
    for mapping in configuration.get(self._QUERY_PARAMETERS_KEY, ())
  File "/Users/swast/src/google-cloud-python/bigquery/google/cloud/bigquery/_helpers.py", line 674, in _query_param_from_api_repr
    return klass.from_api_repr(resource)
  File "/Users/swast/src/google-cloud-python/bigquery/google/cloud/bigquery/_helpers.py", line 440, in from_api_repr
    converted = _CELLDATA_FROM_JSON[type_](value, None)
  File "/Users/swast/src/google-cloud-python/bigquery/google/cloud/bigquery/_helpers.py", line 77, in _datetime_from_json
    return datetime.datetime.strptime(value, _RFC3339_NO_FRACTION)
  File "/Users/swast/.pyenv/versions/3.6.2/lib/python3.6/_strptime.py", line 565, in _strptime_datetime
    tt, fraction = _strptime(data_string, format)
  File "/Users/swast/.pyenv/versions/3.6.2/lib/python3.6/_strptime.py", line 365, in _strptime
    data_string[found.end():])
ValueError: unconverted data remains: .000000
```

Reply:
@tswast The `bigquery-b2` branch has that failure even without this PR.

I'd like to merge here and then fix that branch up (there is also another system test failure due to #3798).

Reply:
In particular, I want to rebase the `bigquery-b2` branch against the current `master` to pick up the subsecond `_datetime_from_json` fix added in #3901.

Reply:
Ack. Let's get the changes from `master` into `bigquery-b2` ASAP.